File: robot-simulator/RobotSimulator.fs
module RobotSimulator

type Direction = North | East | South | West
type Position = int * int

type Robot = { direction: Direction; position: Position }

let create direction position = { direction = direction; position = position }

let turnLeft robot = 
    match robot.direction with
    | North -> { robot with direction = West }
    | East -> { robot with direction = North }
    | South -> { robot with direction = East }
    | West -> { robot with direction = South }

let turnRight robot = 
    match robot.direction with
    | North -> { robot with direction = East }
    | East -> { robot with direction = South }
    | South -> { robot with direction = West }
    | West -> { robot with direction = North }

let advance (robot: Robot) = 
    let (x, y) = robot.position
    match robot.direction with
    | North -> { robot with position = (x, y + 1) }
    | East -> { robot with position = (x + 1, y) }
    | South -> { robot with position = (x, y - 1) }
    | West -> { robot with position = (x - 1, y) }

let instructions (instructions':string) (robot:Robot) = 
    let rec runStep (steps:char[], target:Robot) =
        match Array.tryHead steps with
        | Some 'A' -> runStep (Array.tail steps, advance target)
        | Some 'L' -> runStep (Array.tail steps, turnLeft target)
        | Some 'R' -> runStep (Array.tail steps, turnRight target)
        | _ -> target
    runStep (instructions'.ToCharArray(), robot)

File: csharp/robot-simulator/RobotSimulator.cs
using System;
using System.Linq;

public enum Direction
{
    North, East, South, West
}

public struct Coordinate
{
    public Coordinate(int x, int y) => 
        (X, Y) = (x, y);

    public int X { get; }
    public int Y { get; }
}

public class RobotSimulator
{
    public RobotSimulator(Direction direction, Coordinate coordinate) => 
        (Direction, Coordinate) = (direction, coordinate);

    public Direction Direction { get; private set; }

    public Coordinate Coordinate { get; private set; }

    public void TurnRight()
    {
        var next = (int)Direction + 1;
        Direction = (Direction)(next % 4);
    }

    public void TurnLeft()
    {
        var next = (int)Direction - 1;
        Direction = (Direction)(next < 0 ? 3 : next);
    }

    public void Advance()
    {
        if(Direction == Direction.North)
            Coordinate = new Coordinate(Coordinate.X, Coordinate.Y + 1);
        else if(Direction == Direction.East)
            Coordinate = new Coordinate(Coordinate.X + 1, Coordinate.Y);
        else if(Direction == Direction.South)
            Coordinate = new Coordinate(Coordinate.X, Coordinate.Y - 1);
        else
            Coordinate = new Coordinate(Coordinate.X - 1, Coordinate.Y);
    }

    public void Simulate(string instructions) => 
        instructions.ToList().ForEach(c => 
        {
            if(c == 'A')
                Advance();
            else if(c == 'L')
                TurnLeft();
            else if(c == 'R')
                TurnRight();
            else
                throw new ArgumentException($"An invalid instruction '{c}' was encountered. Instructions can only be 'A', 'L' or 'R'.");
        });
}